net.c

```diff
@@ -7,19 +7,197 @@
+        if (r == -1) {
+            twarn("setting SO_REUSEADDR on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &flags, sizeof flags);
+        if (r == -1) {
+            twarn("setting SO_KEEPALIVE on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, SOL_SOCKET, SO_LINGER, &linger, sizeof linger);
+        if (r == -1) {
+            twarn("setting SO_LINGER on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof flags);
+        if (r == -1) {
+            twarn("setting TCP_NODELAY on fd %d", fd);
+            close(fd);
+            continue;
+        }
+
+        if (verbose) {
+            char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV], *h = host, *p = port;
+            r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                    hbuf, sizeof hbuf,
+                    pbuf, sizeof pbuf,
+                    NI_NUMERICHOST|NI_NUMERICSERV);
+            if (!r) {
+                h = hbuf;
+                p = pbuf;
+            }
+            if (ai->ai_family == AF_INET6) {
+                printf("bind %d [%s]:%s\n", fd, h, p);
+            } else {
+                printf("bind %d %s:%s\n", fd, h, p);
+            }
+        }
+        r = bind(fd, ai->ai_addr, ai->ai_addrlen);
+        if (r == -1) {
+            twarn("bind()");
+            close(fd);
+            continue;
+        }
+
+        r = listen(fd, 1024);
+        if (r == -1) {
+            twarn("listen()");
+            close(fd);
+            continue;
+        }
+
+        break;
+    }
+
+    freeaddrinfo(airoot);
+
+    if(ai == NULL)
+        fd = -1;
+
+    return fd;
+}
+
+static int
+make_unix_socket(char *path)
+{
+    int fd = -1, r;
+    struct stat st;
+    struct sockaddr_un addr;
+    const size_t maxlen = sizeof(addr.sun_path) - 1; // keep room for the null byte
+
+    r = stat(path, &st);
+    if (r == 0) {
+        if (S_ISSOCK(st.st_mode)) {
+            warnx("removing existing local socket to replace it");
+            r = unlink(path);
+            if (r == -1) {
+                twarn("unlink");
+                return -1;
+            }
+        } else {
+            twarnx("another file already exists in the given path");
+            return -1;
+        }
+    }
+
+    fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        twarn("socket()");
+        return -1;
+    }
+
+    r = set_nonblocking(fd);
+    if (r == -1) {
+        close(fd);
+        return -1;
+    }
+
+    if (verbose) {
+        printf("bind %s\n", path);
+    }
+    addr.sun_family = AF_UNIX;
+    if (strlen(path) > maxlen) {
+        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
+        strncpy(addr.sun_path, path, maxlen);
+        addr.sun_path[maxlen] = '\0';
+    } else {
+        strcpy(addr.sun_path, path);
+    }
```

Comment:
There is no need to do copying in the branches. You can leave the warning in the branch and just copy string afterwards. It wont copy more than len of `path` and it will fill the rest of the `addr.sun_path` buffer with zeroes:
```c
    if (strlen(path) > maxlen) {
        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
    }
    strncpy(addr.sun_path, path, maxlen);
```